### DISCUSSION TITLE
Correct orm usage

### DISCUSSION THREAD

**Discussion by amirotin** [sqlalchemy/sqlalchemy#7538] [Usage Questions] [answered]
Hi everyone. I'm trying to understand how to correctly use orm in applications.

For example i have mapped class:

```py
class Account(Base):
    id: int = sa.Column(sa.BigInteger, primary_key=True, autoincrement=False)
    token: str = sa.Column(sa.String(128))
    name: str = sa.Column(sa.String(32))
    level: int = sa.Column(sa.Integer)
    gems: int = sa.Column(sa.Integer)
    vit: int = sa.Column(sa.Integer)
    settings: dict = sa.Column(sa.JSON)
```

At app initialization I init sqlalchemy:
```py
    engine = create_async_engine(
        f"postgresql+asyncpg://{os.getenv('POSTGRES_URL')}", echo=True,
    )

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    session = sessionmaker(
        bind=engine,
        class_=AsyncSession,
        expire_on_commit=False,
    )
```

In order not to duplicate the code I made **DbManager** class where I want to store useful function:
```py
class DbManager:
    def __init__(self, session: sessionmaker):
        self._session: sessionmaker = session

    def __call__(self, **local_kw) -> AsyncSession:
        return self._session(**local_kw)

    async def get_account_id(self, id: int) -> Account:
        async with self._session() as session:
            return await session.get(Account, id)
```

If I get Account (using **get_account_id** function) - how can I update some data during application run?
First option that i see is to store Account instance in application and then call:
```py
        async with self._session() as session:
            session.add(obj)
            await session.commit()
```
Second option is not to store in app and select/update data from account each time it is needed. Then I can make universan update function like:
```py
    async def update_account(self, id, **kwargs):
        async with self._session() as session:
            stmt = sa.update(Account).where(Account.id == id).values(**kwargs)
            await session.execute(stmt)
```
Or I'm completly wrong and there is another way to work with it?

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think the issue is partially because you are creating a session, getting the account, closing the session and then returning the account to the rest of the program. So from your `get_account_id` you get back a detached instance. Refer to https://docs.sqlalchemy.org/en/14/orm/session_state_management.html for the various steps.

Usually a better way of doing this would be to create a session at the beginning of an operation, doing all the queries and then committing/closing the session at the end. This way to change something on an account you can just assign to the instance
```py
Session: sessionmaker = ...
async with Session() as session:
  a = await session.get(Account, id)
  a.name = 'new-name'
  await session.commit()
```

Depending on the type of application you may use some framework to ease the session management.
Like for server like the applications, most framework (think falcon, fastapi, etc) offer a way of creating a session at the beginning of a request, a way of accessing it from the responder function/class and a way of closing it at the end of the session.

Regarding the two function you ways you are trying to update an object, the second one (the one with update) it should work, while you should probably replace `add` with `merge` in the first one. https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.merge
